[JSC] GetFromScope and PutToScope should cache the StructureID

kmiller68 opened the pull request:
#### a512fe6e5cf86987997b9be5f8bbea38fe74c247
<pre>
[JSC] GetFromScope and PutToScope should cache the StructureID
<a href="https://bugs.webkit.org/show_bug.cgi?id=300504">https://bugs.webkit.org/show_bug.cgi?id=300504</a>
<a href="https://rdar.apple.com/162364409">rdar://162364409</a>

Reviewed by Yusuke Suzuki and Dan Hecht.

Right now we the whole Structure* but that&apos;s kinda pointless because we are going to load
a StructureID from the scope we are attempting to load from. Instead we should just cache
the StructureID and compare those directly.

Canonical link: <a href="https://commits.webkit.org/301358@main">https://commits.webkit.org/301358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5189ee2b445cc6478bc88c3f5fabcffebcc7c27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125602 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132464 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77491 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a832e3cc-b80c-4e59-a56e-b69ee211a8a2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45951 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53824 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95680 "17 flakes 23 failures") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63819 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Exiting early after 60 failures. 6458 tests run. 60 failures; Uploaded test results; layout-tests (exception)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cb7819e5-f71e-43e9-94ab-b0baf57e7cef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112318 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76177 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/65994b4d-fea8-453b-8c65-860d136fbde0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35619 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/31986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75937 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117704 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135136 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124113 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52402 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40164 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104151 "44 flakes 79 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103883 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26479 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49238 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27550 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49613 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52292 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58089 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157131 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51644 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39323 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54996 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53339 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1267,12 +1267,12 @@ void JIT::emit_op_get_from_scope(const JSInstruction* currentInstruction)
 
         auto getPutInfoAddress = metadataAddress.withOffset(Metadata::offsetOfGetPutInfo());
-        auto structureAddress = metadataAddress.withOffset(Metadata::offsetOfStructure());
+        auto structureAddress = metadataAddress.withOffset(Metadata::offsetOfStructureID());
```
Maybe rename to structureIDAddress

Constellation replied:
Let's name it as `structureIDAddress`.

dhecht commented on Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
```diff
@@ -2764,7 +2764,7 @@ end)
     get(m_scope, t0)
     loadp PayloadOffset[cfr, t0, 8], t0
-    loadp %opcodeStruct%::Metadata::m_structure[t5], t1
+    loadp %opcodeStruct%::Metadata::m_structureID[t5], t1
```
should this be `loadi`?

kmiller68 replied:
On 32-bit they're the same since the `StructureID` is the `Structure*`. I can change it though.

dhecht replied:
Oh, yeah, maybe p is fine and more correct in that case.

Constellation reviewed (approved):
r=me

dhecht commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -9465,8 +9465,8 @@ void ByteCodeParser::parseBlock(unsigned limit)
                     watchpoints = metadata.m_watchpointSet;
-                else if (resolveType != UnresolvedProperty && resolveType != UnresolvedPropertyWithVarInjectionChecks)
-                    structure = metadata.m_structure.get();
+                else if (resolveType == GlobalProperty || resolveType == UnresolvedPropertyWithVarInjectionChecks)
```
Is this the equivalent condition? i.e. are the other possible resolveType values not possible here?

kmiller68 committed:
[JSC] GetFromScope and PutToScope should cache the StructureID
https://bugs.webkit.org/show_bug.cgi?id=300504
rdar://162364409

Reviewed by Yusuke Suzuki and Dan Hecht.

Right now we the whole Structure* but that's kinda pointless because we are going to load
a StructureID from the scope we are attempting to load from. Instead we should just cache
the StructureID and compare those directly.

Canonical link: https://commits.webkit.org/301358@main
